cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStoreProvider.java

```diff
@@ -0,0 +1,543 @@
+          // We don't want to list the meta data elements separately.
+          if (name.endsWith(METADATA_SUFFIX)) {
+            continue;
+          }
+          list.put(name, getDescription(name));
+        }
+      } catch (KeyStoreException e) {
+        throw new IOException("Failed to get the description for " + name, e);
+      }
+      return list;
+
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  /**
+   * Returns the metadata for the element identified by the given name.
+   * @param name
+   * @return
+   * @throws IOException
+   */
+  SecureStoreMetadata getSecureStoreMetadata(String name) throws IOException {
+    String metaKey = constructMetadataKey(name);
+    if (cache.containsKey(metaKey)) {
+      return cache.get(metaKey);
+    }
+    try {
+      if (!keyStore.containsAlias(metaKey)) {
+        throw new IOException("Metadata for " + name + " not found in the secure store.");
+      }
+      SecureStoreMetadata meta = ((KeyMetadata) keyStore.getKey(metaKey, password)).metadata;
+      cache.put(metaKey, meta);
+      return meta;
+    } catch (NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException e) {
+      throw new IOException("Unable to retrieve the metadata for " + name);
+    }
+  }
+
+  /**
+   * Returns the securely stored data as a UTF8 encoded byte array.
+   * @param name
+   * @return
+   * @throws IOException
+   */
+  byte[] getData(String name) throws IOException {
+    try {
+      if (!keyStore.containsAlias(name)) {
+        throw new IOException(name + " not found in the secure store.");
+      }
+      SecretKeySpec key = (SecretKeySpec) keyStore.getKey(name, password);
+      return key.getEncoded();
+    } catch (NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException e) {
+      throw new IOException("Unable to retrieve the key " + name);
+    }
+  }
+
+  /**
+   * @return the provider configuration
+   */
+  Configuration getcConf() {
+    return cConf;
+  }
+
+  private static Path constructOldPath(Path path) {
+    return Paths.get(path.toString(), "_OLD");
+  }
+
+  private static Path constructNewPath(Path path) {
+    return Paths.get(path.toString(), "_NEW");
+  }
+
+  private static void loadFromPath(KeyStore keyStore, Path path, char[] password)
+    throws IOException {
+    try (InputStream in = new DataInputStream(Files.newInputStream(path))) {
+      keyStore.load(in, password);
+    } catch (NoSuchAlgorithmException | CertificateException e) {
+      throw new IOException("Unable to load the Secure Store. ", e);
+    }
+  }
+
+  private static boolean isBadOrWrongPassword(IOException ioe) {
+    /* According to Java keystore documentation if the load failed due to bad password
+       then the cause of the exception would be set to "UnrecoverableKeyException".
+       Unfortunately that is not the observed behavior. */
+    if (ioe.getCause() instanceof UnrecoverableKeyException) {
+      return true;
+    }
+    // Workaround
```

Comment:
The comment on top explains that. 
According to Java keystore documentation if the load failed due to bad password then the cause of the exception would be set to "UnrecoverableKeyException". Unfortunately that is not the observed behavior. So as a workaround I am looking at the message.